### PR TITLE
feat(e2ei): rotate key packages WPB-2091

### PR DESCRIPTION
**Pull request by David-Henner**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2091" title="WPB-2091" target="_blank">StoryWPB-2091</a>  [iOS] put Certificate in KeyPackages and LeafNodes
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

During the e2ei certificate enrolment flow, we need to re-create key packages including the newly obtained certificate chain.
In this PR we implement the flow of recreating and re-uploading key packages for the self user, as well as updating the existing MLS conversations with the new key packages.

### Changes

- Created action and action handler to replace key packages
- Created `E2eIKeyPackageRotator` that implements the flow of updating the key packages
- Updated the `EnrollE2eiCertificateUseCase` to make use of the key package rotator
- Misc: moved some mocks to `WireDataModelSupport`

### Testing

- [x] I have added automated test to this contribution

### Note

This PR includes some commits cherry-picked from https://github.com/wireapp/wire-ios/pull/787
Once they're merged to develop, I will update this PR

### CODE REVIEW

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        guard let identity = e2eIService.e2eIdentity as? WireE2eIdentity else {
+            throw Error.invalidIdentity
+        }
+
+        // Get the rotate bundle from core crypto
+        let rotateBundle = try await coreCrypto.perform {
+            try await $0.e2eiRotateAll(
+                enrollment: identity,
+                certificateChain: certificateChain,
+                newKeyPackageCount: 100
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift]: move this to property

File: wire-ios-data-model/Source/MLS/Actions/ReplaceSelfMLSKeyPackagesAction.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        case invalidBodyOrCiphersuites
+        case endpointUnavailable
+        case mlsProtocolError
+        case mlsIdentityMismatch
+        case unknown(status: Int)
+
+        public var errorDescription: String? {
+            switch self {
+            case .endpointUnavailable:
+                return "End point not available"
```
Comment (KaterinaWire) [inline wire-ios-data-model/Source/MLS/Actions/ReplaceSelfMLSKeyPackagesAction.swift]: `Endpoint` ?

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift (hunk 1 above)
Comment (KaterinaWire) [reply wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift]: do you want to add this as `TODO`?

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+import Foundation
+import WireCoreCrypto
+import WireDataModel
+
+// sourcery: AutoMockable
+public protocol E2eIKeyPackageRotating {
+
+    func rotateKeysAndMigrateConversations(
+        e2eIService: E2eIServiceInterface,
```
Comment (KaterinaWire) [inline wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift]: I'm wondering if we really need to inject `e2eIService`. We only use the `e2eIdentity` object

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift [hunk 4]
```diff
@@ -271,20 +271,36 @@ public class ZMUserSession: NSObject {
...
             }
+
             let e2eiClient = E2eIClient(coreCrypto: coreCrypto)
-            e2eiRepository = E2eIRepository(acmeApi: acmeApi,
-                                            apiProvider: apiProvider,
-                                            e2eiClient: e2eiClient)
+
+            let keyRotator = E2eIKeyPackageRotator(
+                coreCryptoProvider: coreCryptoProvider,
+                conversationEventProcessor: ConversationEventProcessor(context: syncContext),
```
Comment (KaterinaWire) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift]: I think we don't need to create a new `ConversationEventProcessor`, ZMUserSession has `conversationEventProcessor` property

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift (hunk 3 above)
Comment (David-Henner) [reply wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift]: We don't really need this except for tests

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift (hunk 4 above)
Comment (David-Henner) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift]: nice catch. updated https://github.com/wireapp/wire-ios/pull/793/commits/98fa8015a312a2fe4bb2b34ff638723f2d1a0a15

File: wire-ios-data-model/Source/MLS/Actions/ReplaceSelfMLSKeyPackagesAction.swift (hunk 2 above)
Comment (David-Henner) [reply wire-ios-data-model/Source/MLS/Actions/ReplaceSelfMLSKeyPackagesAction.swift]: fixed https://github.com/wireapp/wire-ios/pull/793/commits/98fa8015a312a2fe4bb2b34ff638723f2d1a0a15

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift (hunk 1 above)
Comment (David-Henner) [reply wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift]: added the property https://github.com/wireapp/wire-ios/pull/793/commits/98fa8015a312a2fe4bb2b34ff638723f2d1a0a15

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotatorTests.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import Foundation
+import XCTest
+import WireCoreCrypto
+
+@testable import WireRequestStrategy
+@testable import WireDataModelSupport
+
+class E2eIKeyPackageRotatorTests: MessagingTestBase {
+
+    private var mockCoreCrypto: MockCoreCrypto!
```
Comment (KaterinaWire) [inline wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotatorTests.swift]: There are new changes with MockCoreCrypto, we should use `MockCoreCryptoProtocol`. That's why tests are failed.

**Comment by KaterinaWire** [comment]
@David-Henner Looks good to me 👍, just needs to fix the tests according to the latest refactoring from the develop branch..

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotatorTests.swift [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        // Core crypto expects the type `WireE2eIdentity`
+        // Unfortunately, we cannot mock or stub this type
+        // So we're unable to test the behaviour of the key package rotator
+        //
+        // If core crypto changes method signatures to expect a type we're able to mock,
+        // Or we implement a wrapper around `WireE2eIdentity` to enable mocking,
+        // Then we can resolve this.
+        // Until then we're unable to test anything that has a dependency on `WireE2eIdentity`
+        //
+        // TODO: Create a ticket to track this
```
Comment (caldrian) [inline wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotatorTests.swift]: Please add the ticket number.

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift [hunk 7]
```diff
@@ -81,6 +81,9 @@ public final class EnrollE2eICertificateUseCase: EnrollE2eICertificateUseCaseInt
         let finalizeResponse = try await enrollment.finalize(location: orderResponse.location, prevNonce: orderResponse.acmeResponse.nonce)
         let certificateRequest = try await enrollment.certificateRequest(location: finalizeResponse.location, prevNonce: finalizeResponse.acmeResponse.nonce)
 
+        // TODO: Pass the certificate chain
```
Comment (caldrian) [inline wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: When can the TODO be resolved?

**Review by caldrian — APPROVED** [review]

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotatorTests.swift (hunk 6 above)
Comment (David-Henner) [reply wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotatorTests.swift]: done https://github.com/wireapp/wire-ios/pull/793/commits/4b8d7c3ddab7072717ee68f88f6ace271aef77c8

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift (hunk 7 above)
Comment (David-Henner) [reply wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: @KaterinaWire we should get the certificate chain in an upcoming ticket right? Do we have a ticket number I could add to the todo?

Comment (KaterinaWire) [reply wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: You can pass `certificateRequest.response` converted to a string.
I also created [this ticket](https://wearezeta.atlassian.net/browse/WPB-6039) as a reminder to verify enrollment flow after the CC bump.

**Review by KaterinaWire — APPROVED** [review]

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift (hunk 7 above)
Comment (David-Henner) [reply wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: covered here https://github.com/wireapp/wire-ios/pull/793/commits/eb85a1647dcc5132a9c92e895d52b01ae272ce28

**Pull request merged**